src/statedynamics.jl

```diff
@@ -0,0 +1,225 @@
+
+# code for state-space systems
+module StateDynamics
+
+
+
+##############################################################################
+# asbtract
+
+abstract type StateSystem
+end
+
+##############################################################################
+# PI controller
+
+mutable struct PIStateSystem <: StateSystem
+    x
+    kp
+    ki
+end
+
+function PIStateSystem(kp, ki)
+    return PIStateSystem(0, kp, ki)
+end
+
+function next(K::PIStateSystem, u)
+    next_state =  K.x + u 
+    y =  K.kp * u +  K.ki * next_state
+    K.x = next_state
+    return y
+end
+
+
+##############################################################################
+# P controller
+
+
+mutable struct StaticStateSystem <: StateSystem
+    f
+end
+
+
+next(K::StaticStateSystem, u) = K.f(u)
+
+##############################################################################
+
+mutable struct StaticPlusConstantStateSystem <: StateSystem
+    f
+    constant
+end
+
+
+next(K::StaticPlusConstantStateSystem, u) = K.f(u) + K.constant
+
+function setconstant(K::StaticPlusConstantStateSystem, c)
+    K.constant = c
+end
+
+
+##############################################################################
+# Constant controller
+
+mutable struct ConstantStateSystem <: StateSystem
+    constant
+end
+
+
+next(K::ConstantStateSystem, u) =  K.constant
+
+function setconstant(K::ConstantStateSystem, c)
+    K.constant = c
+end
+
+
+##############################################################################
+# switching wrapper
+
+mutable struct SequentialStateSystem
+    count        # time
+    lastoutput
+    subsystems   # list of controller objects
+    endtimes     # times on which to switch
+    switches     # list of functions called at switch events
+
+end
+
+function SequentialStateSystem(subsystems, endtimes, switches)
+    return SequentialStateSystem(0, 0, subsystems, endtimes, switches)
+end
+
+# endtimes = [5, 10]
+#   which_subsystem = 1  for t <= 5
+#   which_subsystem = 2  for 6 <= t <= 10
+which_subsystem(endtimes, t) = searchsortedfirst(endtimes, t)
+
+
+# controller t takes state t and measurement, determines state t+1
+function next(K::SequentialStateSystem, measurement)
+    t = K.count 
+
+    previous_subsystem_index = which_subsystem(K.endtimes, t-1)
+    this_subsystem_index = which_subsystem(K.endtimes, t)
+    if this_subsystem_index > previous_subsystem_index
```

Comment:
The index comparison here implies that subsystems are monotonically increasing with time, and thus, repeating patterns need to be unrolled. Even if this is sufficient for the current set of simulated systems, it may be worth adding a comment explaining the design in the SequentialStateSystem struct.